### PR TITLE
Multiple keymanager support UI

### PR DESCRIPTION
**Pull request by tharindu1st**

### CODE REVIEW

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/ClaimMapping.jsx [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+        const newMapping = claimMappings.filter(({ remoteClaim }) => remoteClaim !== claimKey);
+        setClaimMapping(newMapping);
+    };
+    return (
+        <Box mb={3}>
+            <TableContainer component={Paper}>
+                <Table className={classes.table} aria-label='simple table'>
+                    <TableHead>
+                        <TableRow>
+                            <TableCell>Remote Claim</TableCell>
```
Comment (kavishkafernando) [inline features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/ClaimMapping.jsx:99]: Shall we add i18n to all static values?

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/ClaimMapping.jsx [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+                                    onChange={onChange}
+                                    error={validationError.remoteClaim}
+                                    helperText={validationError.remoteClaim && validationError.remoteClaim}
+                                    value={newRemoteClaim === null ? '' : newRemoteClaim}
+                                />
+                            </TableCell>
+                            <TableCell align='right'>
+                                <TextField
+                                    id='localClaim'
+                                    label='Local Claim'
```
Comment (kavishkafernando) [inline features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/ClaimMapping.jsx:120]: Add i18n for Textfield labels as well

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/ClaimMapping.jsx [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+                                </Grid>
+                            </TableCell>
+                        </TableRow>
+                        {claimMappings.map(({ remoteClaim, localClaim }) => (
+                            <TableRow key={localClaim}>
+                                <TableCell component='th' scope='row'>
+                                    <TextField
+                                        id={remoteClaim}
+                                        defaultValue={remoteClaim}
+                                        label='Remote Claim'
```
Comment (kavishkafernando) [inline features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/ClaimMapping.jsx:157]: add i18n

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/ClaimMapping.jsx [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+                                        variant='outlined'
+                                        disabled
+                                    />
+                                </TableCell>
+                                <TableCell align='right'>
+                                    <TextField
+                                        id={localClaim}
+                                        defaultValue={localClaim}
+                                        label='Local Claim'
```
Comment (kavishkafernando) [inline features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/ClaimMapping.jsx:166]: Add i18n

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/KeyValidation.jsx [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+
+const useStyles = makeStyles((theme) => ({
+    mandatoryStar: {
+        color: theme.palette.error.main,
+        marginLeft: theme.spacing(0.1),
+    },
+}));
+
+/**
+ * Trottling Policies dropdown selector used in minimized API Create form
```
Comment (kavishkafernando) [inline features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/KeyValidation.jsx:34]: Typoe : Throttling

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/KeyValidation.jsx [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+                body[jwtValue.claimKey] = jwtValue.claimValueRegex;
+                tokenValidation.value.body = body;
+                setTokenValidation(tokenValidation);
+            }
+            clearValues();
+            return true;
+        }
+    };
+    const onChange = (e) => {
+        const { value } = e.target;
```
Comment (kavishkafernando) [inline features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/KeyValidation.jsx:90]: we can deconstruct e.target.name as well

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/KeyValidation.jsx [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+ * Trottling Policies dropdown selector used in minimized API Create form
+ * @export
+ * @param {*} props
+ * @returns {React.Component}
+ */
+export default function KeyValidation(props) {
+    const classes = useStyles();
+
+    const { tokenValidation, setTokenValidation } = props;
+    const [, setValidationError] = useState([]);
```
Comment (kavishkafernando) [inline features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/KeyValidation.jsx:43]: validationError? Are we not using this variable but only setting a value?

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/KeyValidation.jsx [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+                />
+            );
+        } else if (type === 'JWT') {
+            return (
+                <Box mb={3}>
+                    <TableContainer component={Paper}>
+                        <Table className={classes.table} aria-label='simple table'>
+                            <TableHead>
+                                <TableRow>
+                                    <TableCell>Claim Key</TableCell>
```
Comment (kavishkafernando) [inline features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/KeyValidation.jsx:135]: Add i18n

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/KeyValidation.jsx [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+                                            label='Claim Key'
+                                            variant='outlined'
+                                            onChange={onCreateJWTmapping}
+                                            value={jwtValue.claimKey}
+                                        />
+                                    </TableCell>
+                                    <TableCell align='right'>
+                                        <TextField
+                                            name='claimValueRegex'
+                                            label='Claim Value Regex'
```
Comment (kavishkafernando) [inline features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/KeyValidation.jsx:154]: Lets add 18n to all static Table cells and Text fields

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/KeyValidation.jsx [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+            </ExpansionPanelSummary>
+            <Box display='flex' flexDirection='row'>
+                <FormControl variant='outlined'>
+                    <Select
+                        name='type'
+                        value={tokenValidation.type}
+                        onChange={onChange}
+                        classes={{ root: classes.slectRoot }}
+                    >
+                        <MenuItem value='REFERENCE'>REFERENCE</MenuItem>
```
Comment (kavishkafernando) [inline features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/KeyValidation.jsx:257]: Add i18n for these as well

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/KeyValidations.jsx [hunk 11]
```diff
@@ -0,0 +1,32 @@
+import React from 'react';
+import sortby from 'lodash.sortby';
+import KeyValidation from './KeyValidation';
+
+
+/**
+ * Trottling Policies dropdown selector used in minimized API Create form
```
Comment (kavishkafernando) [inline features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/KeyValidations.jsx:7]: Typo : Throttling

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/ListKeyManagers.jsx [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+                />
+            </Button>
+        </RouterLink>
+    );
+    const emptyBoxProps = {
+        content: (
+            <Typography variant='body2' color='textSecondary' component='p'>
+                <FormattedMessage
+                    id='AdminPages.KeyManagers.List.empty.content.keymanagers'
+                    defaultMessage='It is possible to register Oauth Provider.'
```
Comment (kavishkafernando) [inline features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/ListKeyManagers.jsx:123]: 'It is possible to register **an** Oauth Provider'

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/SelectKeyManager.jsx [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+
+const useStyles = makeStyles((theme) => ({
+    mandatoryStar: {
+        color: theme.palette.error.main,
+        marginLeft: theme.spacing(0.1),
+    },
+}));
+
+/**
+ * Trottling Policies dropdown selector used in minimized API Create form
```
Comment (kavishkafernando) [inline features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/SelectKeyManager.jsx:20]: Type: Throttling

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/data/api.js [hunk 14]
```diff
@@ -583,6 +583,57 @@ class API extends Resource {
             }),
         );
     }
+    getKeyManagersList() {
```
Comment (kavishkafernando) [inline features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/data/api.js:586]: Shall we add a method comment to this?

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/Base/RouteMenuMapping.jsx [hunk 15]
```diff
@@ -106,6 +107,15 @@ const RouteMenuMapping = (intl) => [
         component: () => <MicrogatewayLabels />,
         icon: <PhonelinkSetupIcon />,
     },
+    {
+        id: intl.formatMessage({
+            id: 'Base.RouteMenuMapping.keymanagers',
+            defaultMessage: 'KeyManagers',
+        }),
+        path: '/settings/key-managers',
+        component: () => <KeyManagers />,
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/Base/RouteMenuMapping.jsx:116]: In the component prop , we can just pass the component variable, no need to wrap it with a function(lambda)

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/ClaimMapping.jsx [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+        if (e.target.id === 'remoteClaim') {
+            setRemoteClaim(e.target.value);
+        } else if (e.target.id === 'localClaim') {
+            setLocalClaim(e.target.value);
+        }
+    };
+    const validate = (fieldName, value) => {
+        let error = '';
+        if (value === null || value === '') {
+            error = 'Calim is Empty';
```
[the remark this answers is not in the document]
Comment (tmkasun) [reply features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/ClaimMapping.jsx:49]: And I think we need to support i18n in these error messages

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/ClaimMapping.jsx [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+    const handleAddToList = () => {
+        const remoteClaimError = validate('remoteClaim', newRemoteClaim);
+        const localClaimError = validate('localClaim', newLocalClaim);
+        if (localClaimError !== '' || remoteClaimError !== '') {
+            Alert.error(remoteClaimError);
+            clearValues();
+            return false;
+        } else {
+            let exist = false;
+            claimMappings.map(({ remoteClaim, localClaim }) => {
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/ClaimMapping.jsx:70]: I think here we are looking for at least one `exist`?
If so we can use `for of` loop to iterate the `claimMappings` and break on first `exist`

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/KeyManagerConfiguration.jsx [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+ * Trottling Policies dropdown selector used in minimized API Create form
+ * @export
+ * @param {*} props
+ * @returns {React.Component}
+ */
+export default function KeyManagerConfiguration(props) {
+    const { keymanagerConnectorConfigurations, additionalProperties, setAdditionalProperties } = props;
+    const onChange = (e) => {
+        let { value } = e.target;
+        if (e.target.type === 'checkbox') {
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/KeyManagerConfiguration.jsx:23]: Please destructure the type as well

i:e
```javascript
let { value, type } = e.currentTarget;
```

Comment (tmkasun) [reply features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/KeyManagerConfiguration.jsx:23]: and `name` too

**Comment by tmkasun** [comment]
@tharindu1st ,

This is impressive 😃 
Seems like you have fallen 💘 with JavaScript 😄

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/SelectKeyManager.jsx [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+                    </>
+                )}
+                value={keyManager}
+                name='keyManager'
+                onChange={handleValidateAndChange}
+                SelectProps={{
+                    multiple,
+                    renderValue: (selected) => (Array.isArray(selected) ? selected.join(', ') : selected),
+                }}
+                helperText={isAPIProduct ? helperText + 'API Product' : helperText + 'API'}
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/KeyManagers/SelectKeyManager.jsx:58]: Is this helper text correct here ? `API Product` ?

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/Base/RouteMenuMapping.jsx (hunk 15 above)
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/Base/RouteMenuMapping.jsx:116]:
```suggestion
        component: KeyManagers,
```

**Pull request merged**